src/packet.ts

```diff
@@ -268,180 +272,172 @@ export class ConnectionAssetDetailsFrame extends BaseFrame {
     this.errorMessage = errorMessage
   }
 
   static fromContents (reader: Reader): StreamCloseFrame {
-    const streamId = reader.readVarUIntBigNum()
-    const errorCode = reader.readUInt8BigNum().toNumber() as ErrorCode
+    const streamId = reader.readVarUIntLong()
+    const errorCode = reader.readUInt8Number() as ErrorCode
     const errorMessage = reader.readVarOctetString().toString('utf8')
     return new StreamCloseFrame(streamId, errorCode, errorMessage)
   }
 }
 
 export class StreamMoneyFrame extends BaseFrame {
   type: FrameType.StreamMoney
-  streamId: BigNumber
-  shares: BigNumber
+  streamId: Long
+  shares: Long
 
-  constructor (streamId: BigNumber.Value, shares: BigNumber.Value) {
+  constructor (streamId: LongValue, shares: LongValue) {
     super('StreamMoney')
-    this.streamId = new BigNumber(streamId)
-    this.shares = new BigNumber(shares)
-
-    assert(this.shares.isInteger() && this.shares.isPositive(), `shares must be a positive integer: ${shares}`)
+    this.streamId = longFromValue(streamId, true)
+    this.shares = longFromValue(shares, true)
   }
 
   static fromContents (reader: Reader): StreamMoneyFrame {
-    const streamId = reader.readVarUIntBigNum()
-    const amount = reader.readVarUIntBigNum()
+    const streamId = reader.readVarUIntLong()
+    const amount = reader.readVarUIntLong()
     return new StreamMoneyFrame(streamId, amount)
   }
 }
 
 export class StreamMaxMoneyFrame extends BaseFrame {
   type: FrameType.StreamMaxMoney
-  streamId: BigNumber
-  receiveMax: BigNumber
-  totalReceived: BigNumber
+  streamId: Long
+  receiveMax: Long
+  totalReceived: Long
 
-  constructor (streamId: BigNumber.Value, receiveMax: BigNumber.Value, totalReceived: BigNumber.Value) {
+  constructor (streamId: LongValue, receiveMax: LongValue, totalReceived: LongValue) {
     super('StreamMaxMoney')
-    this.streamId = new BigNumber(streamId)
-    this.receiveMax = new BigNumber(receiveMax)
-    this.totalReceived = new BigNumber(totalReceived)
-
-    if (!this.receiveMax.isFinite()) {
-      this.receiveMax = MAX_UINT64
+    if (typeof receiveMax === 'number' && !isFinite(receiveMax)) {
+      receiveMax = Long.MAX_UNSIGNED_VALUE
     }
 
-    assert(this.receiveMax.isInteger() && this.receiveMax.isPositive(), `receiveMax must be a positive integer. got: ${receiveMax}`)
-    assert(this.totalReceived.isInteger() && this.totalReceived.isPositive(), `totalReceived must be a positive integer. got: ${totalReceived}`)
+    this.streamId = longFromValue(streamId, true)
+    this.receiveMax = longFromValue(receiveMax, true)
+    this.totalReceived = longFromValue(totalReceived, true)
   }
 
   static fromContents (reader: Reader): StreamMaxMoneyFrame {
-    const streamId = reader.readVarUIntBigNum()
-    const receiveMax = reader.readVarUIntBigNum()
-    const totalReceived = reader.readVarUIntBigNum()
+    const streamId = reader.readVarUIntLong()
+    const receiveMax = reader.readVarUIntLong()
```

Comment:
I think we need to be careful here if we're going to roll this out. SPSP servers are typically willing to accept an infinite amount of money, but they often express it by setting receiveMax to a very large number. If that number doesn't fit in 8 bytes, then the SPSP server becomes unpayable with this new version.

I think there a few ways to deal with this:

1. We could look through all implementations and ensure they're all upgraded.
2. We could store a special value if the receiveMax is >8 bytes, pay up until that point, and then sever when we reach the limit
3. If the receiveMax is >8 bytes just store it as `2 ** 64 - 1` so that it will never be exceeded. If sendMax is ever >8 bytes, however, throw an error. That way the sendMax will never exceed the receiveMax but we also don't need to represent a >8 byte receiveMax. We could do it the other way around if the receiver is on the new version. Represent a sendMax of >8 bytes as the max 8 byte value and forbid the receiveMax from being >8 bytes.

Thoughts?

Reply:
@sharafian I like 3 best -- easier upgrade path than 1, and I think it'll have a more natural implementation than 2.